### PR TITLE
(fix): Configure select concept answers to search based on referenced field

### PR DESCRIPTION
**Pull request by arodidev**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

Within AFE, the functionality for `select-concept-answers` is that the datasource is configured make a network request using the field's `conceptUuid`. However, in the event that an answer is selected within an associated field, the datasource will make a network request based on the uuid tied to the answer selected in the associated field, thus giving a different list in the dropdown with every different answer selected.
 
This PR adds this functionality, enabling the dropdown to be populated based on a the answer selected in a referenced field.

It also enables `select-concept-answers`'s list of answers to be populated by data from a `Coded` concept. 

Sample schema: 
```json
{
  "name": "Test select concept answers",
  "pages": [
    {
      "label": "page one",
      "sections": [
        {
          "label": "First Section",
          "isExpanded": "true",
          "questions": [
            {
              "label": "Ped WHO stage",
              "type": "obs",
              "id": "pedWhoStage",
              "questionOptions": {
                "concept": "a89b3740-1350-11df-a1f1-0026b9348838",
                "answers": [
                  {
                    "label": "1",
                    "concept": "162552AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
                  },
                  {
                    "label": "2",
                    "concept": "2462a9d7-61fb-4bf5-9359-aedecb8d03cb"
                  },
                  {
                    "label": "3",
                    "concept": "167157AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
                  },
                  {
                    "label": "4",
                    "concept": "0eaa21e0-5f69-41a9-9dea-942796590bbb"
                  }
                ],
                "rendering": "select"
              },
              "validators": [
                {
                  "type": "js_expression",
                  "failsWhenExpression": "isEmpty(myValue) && arrayContains (['18faa058-4eea-4339-a959-84b3e5cb30be', '03f5b165-c577-4e1e-b7ae-acfc3ba18ebf'], visitTypeUuid)",
                  "message": "Required."
                }
              ]
            },
            {
              "label": "Criteria",
              "id": "criteria",
              "type": "obs",
              "questionOptions": {
                "concept": "162552AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                "rendering": "select-concept-answers",
                "config":{
                  "referencedField":"pedWhoStage"
                }
              }
            },
             {
              "label": " New WHO staging:",
              "type": "obs",
              "required": "false",
              "questionOptions": {
                "concept": "a8a084ca-1350-11df-a1f1-0026b9348838",
                "answers": [
                  {
                    "concept": "a899b35c-1350-11df-a1f1-0026b9348838",
                    "label": "Yes"
                  },
                  {
                    "concept": "a899b42e-1350-11df-a1f1-0026b9348838",
                    "label": "No"
                  }
                ],
                "rendering": "select"
              },
              "id": "__FrEGLuv3v"
            }
          ]
        }
      ]
    }
  ],
  "processor": "EncounterFormProcessor",
  "encounterType": "",
  "referencedForms": [],
  "uuid": "",
  "description": "Sample form for testing select concept answers"
}
```

## Screenshots

https://github.com/openmrs/openmrs-form-engine-lib/assets/51090527/ec7d311b-5924-49f2-99c6-f835e7564cd5

### CODE REVIEW

**Comment by samuelmale** [comment]
> However, in the event that an answer is selected within an associated field, the datasource will make a network request based on the uuid tied to the answer selected in the associated field, thus giving a different list in the dropdown with every different answer selected.

@arodidev how is this kind of association or dependency configured in the form?

**Comment by arodidev** [comment]
>how is this kind of association or dependency configured in the form?

The form designer would need to add an additional property to the schema which references the associated field. I should add a sample schema to the summary for context, my bad.

File: src/datasources/concept-set-and-answers-data-source.ts [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-    super(`${restBaseUrl}/concept/conceptUuid?v=custom:(uuid,setMembers:(uuid,display))`);
+    super(`${restBaseUrl}/concept/conceptUuid?v=custom:(uuid,setMembers:(uuid,display),answers:(uuid,display))`);
   }
 
   fetchData(searchTerm: string, config?: Record<string, any>): Promise<any[]> {
-    const apiUrl = this.url.replace('conceptUuid', config.concept);
+    const apiUrl = this.url.replace('conceptUuid', config.value || config.concept);
     return openmrsFetch(apiUrl).then(({ data }) => {
-      return data['setMembers'];
+      return data['setMembers']?.length ? data['setMembers'] : data['answers'];
```
[the remark this answers is not in the document]
Comment (arodidev) [reply src/datasources/concept-set-and-answers-data-source.ts:12]: @usamaidrsk yes, AFE currently allows only `setMembers` to populate the dropdown. There was a request to expand this functionality to allow populating using `Coded` concept as well.

**Comment by usamaidrsk** [comment]
@arodidev will it be possible to validate the `referencedField` that it's answer will be a `Coded` concept, I am asking in relation to the schema validations on the form builder.

File: src/datasources/concept-set-and-answers-data-source.ts [hunk 2]
```diff
@@ -1,15 +1,15 @@
 import { openmrsFetch, restBaseUrl } from '@openmrs/esm-framework';
 import { BaseOpenMRSDataSource } from './data-source';
 
-export class ConceptSetMembersDataSource extends BaseOpenMRSDataSource {
+export class ConceptSetAndAnswersDataSource extends BaseOpenMRSDataSource {
```
Comment (eudson) [inline src/datasources/concept-set-and-answers-data-source.ts:4]: @arodidev why not calling it `SelectConceptAnswersDataSource`?

**Comment by arodidev** [comment]
will it be possible to validate the `referencedField` that it's answer will be a `Coded` concept, I am asking in relation to the schema validations on the form builder.

With the current implementation, the referenced field that is updating `select-concept-answers` will be a `dropdown` which will always have `Coded` (unless form builder validation concept validation determines otherwise). This you could consider to be some sort of implicit validation that the referenced field will always have `Coded` concepts.

File: src/datasources/concept-set-and-answers-data-source.ts (hunk 2 above)
Comment (arodidev) [reply src/datasources/concept-set-and-answers-data-source.ts:4]: Thanks @eudson, will resolve.

**Comment by arodidev** [comment]
I am closing this PR to refactor this logic in a different PR

**Pull request closed without merging**